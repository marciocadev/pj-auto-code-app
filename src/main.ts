import { join } from 'path';
import { App, Aspects, Stack, StackProps } from 'aws-cdk-lib';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { AwsSolutionsChecks, NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';
import { UserTable, GrantType } from './constructs/user-table';

/**
 * Basic stack
 */
export class PjAutoCodeAppStack extends Stack {
  /**
   * Create a new PjAutoCodeAppStack
   *
   * @param scope - scope in which this resource is defined
   * @param id    - scoped id of the resource
   * @param props - resource properties
   */
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    const users = new UserTable(this, 'UserTable');

    const createHandler = new NodejsFunction(this, 'CreateUser', {
      entry: join(__dirname + '/lambda-fns/create-user.ts'),
      handler: 'handler',
      runtime: Runtime.NODEJS_14_X,
      bundling: {
        minify: true,
      },
    });
    users.bind(createHandler, GrantType.Write);
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

const stack = new PjAutoCodeAppStack(app, 'my-stack-dev', { env: devEnv });
// new MyStack(app, 'my-stack-prod', { env: prodEnv });

Aspects.of(app).add(new AwsSolutionsChecks());
NagSuppressions.addStackSuppressions(stack, [
  { id: 'AwsSolutions-IAM4', reason: 'handler receive the correct table grant to action' },
]);

app.synth();