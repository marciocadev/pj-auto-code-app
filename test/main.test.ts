import { App } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { PjAutoCodeAppStack } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new PjAutoCodeAppStack(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});

describe(('validate my stack'), () => {
  let app: App;
  let stack: PjAutoCodeAppStack;
  let template: Template;

  beforeAll(() => {
    // GIVEN
    app = new App();
    // WHEN
    stack = new PjAutoCodeAppStack(app, 'test');
    // THEN
    template = Template.fromStack(stack);
  });

  test(('count tables'), () => {
    template.resourceCountIs('AWS::DynamoDB::Table', 1);
  });

  test(('varify partition and sort keys'), () => {
    template.hasResourceProperties('AWS::DynamoDB::Table', {
      KeySchema: [
        Match.objectLike({
          AttributeName: 'username',
          KeyType: 'HASH',
        }),
        Match.objectLike({
          AttributeName: 'loginDate',
          KeyType: 'RANGE',
        }),
      ],
    });
  });

  test(('verify table attributes'), () => {
    template.hasResourceProperties('AWS::DynamoDB::Table', {
      AttributeDefinitions: [
        Match.objectLike({
          AttributeName: 'username',
          AttributeType: 'S',
        }),
        Match.objectLike({
          AttributeName: 'loginDate',
          AttributeType: 'S',
        }),
      ],
    });
  });

  test(('verify table removal policy'), () => {
    template.hasResource('AWS::DynamoDB::Table', {
      UpdateReplacePolicy: Match.exact('Delete'),
    });
  });

  test('count lambdas', () => {
    template.resourceCountIs('AWS::Lambda::Function', 4);
  });

});