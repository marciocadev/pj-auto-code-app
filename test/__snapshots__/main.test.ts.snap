// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "CreateUser1C98808B": Object {
      "DependsOn": Array [
        "CreateUserServiceRoleDefaultPolicy44CE0BDC",
        "CreateUserServiceRole583944EB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "de104f3e313e99087fca8aad1c5964d05ec06a7641c933ef9d6085b7f9f5fde2.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "USER_TABLE_NAME": Object {
              "Ref": "UserTableBD4BF69E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CreateUserServiceRole583944EB",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "CreateUserServiceRole583944EB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CreateUserServiceRoleDefaultPolicy44CE0BDC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "UserTableBD4BF69E",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CreateUserServiceRoleDefaultPolicy44CE0BDC",
        "Roles": Array [
          Object {
            "Ref": "CreateUserServiceRole583944EB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GetUserC5B494A2": Object {
      "DependsOn": Array [
        "GetUserServiceRoleDefaultPolicy148A771F",
        "GetUserServiceRoleA2CA9640",
      ],
    "DeleteUser55A9FE77": Object {
      "DependsOn": Array [
        "DeleteUserServiceRoleDefaultPolicyA88C2C85",
        "DeleteUserServiceRoleF9B1C57A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f415d368a9931a8eba266081cea9b38a58c8c17fe4f7c2cd1181aac9b8fa2e6a.zip",
          "S3Key": "0fa97cc4e336aa1241f07736d927d846f9d84a4883781f7c242437bba82dd112.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "USER_TABLE_NAME": Object {
              "Ref": "UserTableBD4BF69E",
            },
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "GetUserServiceRoleA2CA9640",
            "DeleteUserServiceRoleF9B1C57A",

              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },

    "UserTableBD4BF69E": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "username",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "code",
            "AttributeType": "N",
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "username",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "code",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": Object {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
